pkg/function/service.go

```diff
@@ -230,15 +215,42 @@ func (fs *Service) ReduceFn(stream functionpb.UserDefinedFunction_ReduceFnServer
 			eventTime: d.GetEventTime().EventTime.AsTime(),
 			watermark: d.GetWatermark().Watermark.AsTime(),
 		}
-		reduceCh <- hd
+
+		ch, chok := chanMap[d.Key]
+		if !chok {
+			ch = make(chan Datum)
+			chanMap[d.Key] = ch
+
+			wg.Add(1)
+			go func(key string, ch chan Datum) {
+				defer wg.Done()
+				messages := fs.Reducer.HandleDo(ctx, key, ch, md)
+				mu.Lock()
+				defer mu.Unlock()
+				for _, msg := range messages {
+					datumList = append(datumList, &functionpb.Datum{
```

Comment:
Instead of using a lock, would it be better to use a channel for the returned datum?

Reply:
We need to rethink the return https://github.com/numaproj/numaflow/issues/509